### PR TITLE
Set up CI for aarch64-apple-darwin

### PR DESCRIPTION
**Pull request by shepmaster**

### CODE REVIEW

**Comment by shepmaster** [comment]
@pietroalbini 

[`f9d308e` (#75991)](https://github.com/rust-lang/rust/pull/75991/commits/f9d308e92d5724163c7771a9fb0b7f1e54212321) is working. There are a lot of other commits that correspond to other PRs that need to go in first, so I'd only pay attention to that one commit.

It has some Azure gunk in it still, which I'll remove. Beyond that, how do the changes look?

**Comment by shepmaster** [comment]
@pietroalbini FWIW, I'm rebasing this as the other required PRs get merged, so the previous SHA isn't the useful one to check. You'll want to look for the commit titled "Add a cross-compiling aarch64-apple-darwin CI builder".

File: src/ci/github-actions/ci.yml
Comment (pietroalbini) [inline src/ci/github-actions/ci.yml:315]: This is not needed: I just checked with the DTK, and the `xcrun --sdk macosx --show-sdk-path` we're doing in `install-clang.sh` returns the correct SDK root.

File: src/ci/github-actions/ci.yml
Comment (pietroalbini) [inline src/ci/github-actions/ci.yml:156]: This step needs to be above `install-clang.sh` to make `xcrun --sdk macosx --show-sdk-path` work.

File: src/ci/github-actions/ci.yml
Comment (pietroalbini) [inline src/ci/github-actions/ci.yml:158]: When we switched to GHA we migrated all the setup code to standalone bash scripts in `src/ci/scripts/{name}.sh`. Could you move both the command and the `SELECT_XCODE`  condition to it? A comment on top of the script to explain why that's needed would also be great!

File: src/ci/scripts/install-clang.sh
Comment (pietroalbini) [inline src/ci/scripts/install-clang.sh:19]: Is this intended to be temporary until we bump the LLVM version used in CI to one supporting macOS 11?

File: silicon/README.md
Comment (pietroalbini) [inline silicon/README.md:1]: Can we move the setup instructions to the [platform specific instructions](https://rustc-dev-guide.rust-lang.org/building/prerequisites.html#platform-specific-instructions) section of the rustc-dev-guide? I'd really prefer not to have another top-level directory.

**Review by emilyalbini — COMMENTED** [review]
Thanks for working on this! Left some comments.

File: src/ci/github-actions/ci.yml
Comment (shepmaster) [reply src/ci/github-actions/ci.yml:315]: It’s needed because we aren’t running this builder on a DTK.

File: src/ci/github-actions/ci.yml
Comment (shepmaster) [reply src/ci/github-actions/ci.yml:156]: That line is disabled because it’s passed in via the environment variable.

File: silicon/README.md
Comment (shepmaster) [reply silicon/README.md:1]: These instructions will not be merged, that’s why I suggested you only review the single commit 😇

File: src/ci/github-actions/ci.yml
Comment (pietroalbini) [reply src/ci/github-actions/ci.yml:315]: That should still work. I did a test in a private repository of mine, and that command shows the 11.0 SDK path:

Screenshot from 2020-09-21 14-08-45

File: src/ci/github-actions/ci.yml
Comment (pietroalbini) [reply src/ci/github-actions/ci.yml:156]: Yep. Still, running it beforehand could prevent surprising behavior in the future.

File: silicon/README.md
Comment (pietroalbini) [reply silicon/README.md:1]: Heh.

File: src/ci/github-actions/ci.yml
Comment (shepmaster) [reply src/ci/github-actions/ci.yml:315]: Sure, I can re-enable that line of code.

File: src/ci/github-actions/ci.yml
Comment (shepmaster) [reply src/ci/github-actions/ci.yml]: Instead, I just merged it with `install-clang` since the two were so intertwined anyway. That also removes the ordering issue.

File: src/ci/scripts/install-clang.sh
Comment (shepmaster) [reply src/ci/scripts/install-clang.sh]: I don't know, actually! That seems like it would be nice.

Comment (pietroalbini) [reply src/ci/scripts/install-clang.sh]: I think it's really useful to converge macOS AArch64 to use the same upstream version as the other platforms, as it reduces the chance of encountering Apple Silicon-only bugs. I totally understand that's not the case today, but do you know if we expect it to be possible to use the next upstream version of LLVM instead of the one bundled with Xcode?

It'd be nice to have a comment on this in the source too!

File: src/ci/scripts/install-clang.sh
Comment (pietroalbini) [inline src/ci/scripts/install-clang.sh:18]: I'd split this in two independent environment variables, each with its own separate `if` statement: `SELECT_XCODE` to choose a custom Xcode version, and `USE_XCODE_CLANG` to use the bundled clang instead of our own. This will allow to reuse `SELECT_XCODE` in the future even in builders that can use the native clang.

File: src/ci/scripts/install-clang.sh
Comment (pietroalbini) [inline src/ci/scripts/install-clang.sh:19]: Following up on the "split this into two environment variables", we can avoid relying on `SELECT_XCODE` to find clang:

```suggestion
        bindir="$(xcode-select --print-path)/Toolchains/XcodeDefault.xctoolchain/usr/bin"
```

**Review by emilyalbini — COMMENTED** [review]
Another round of review, I feel this is production ready once this is addressed!

File: src/ci/scripts/install-clang.sh
Comment (shepmaster) [reply src/ci/scripts/install-clang.sh:18]: > I'd split this in two independent environment variables

In that case, do you want these split back into two separate steps and scripts?

Comment (pietroalbini) [reply src/ci/scripts/install-clang.sh]: Possibly, yeah.

**Comment by pietroalbini** [comment]
Changes look good by the way!

**Comment by shepmaster** [comment]
> Changes look good by the way!

Or they would if I knew how booleans worked 🤦‍♂️

**Comment by shepmaster** [comment]
Ok, with that try build I can do something like

```
% rustup-toolchain-install-master --force 462ec6baef52edea591eab1a070ed4e18b80d094 --targets aarch64-apple-darwin

% SDKROOT=$(xcrun -sdk macosx11.0 --show-sdk-path) MACOSX_DEPLOYMENT_TARGET=$(xcrun -sdk macosx11.0 --show-sdk-platform-version) cargo +462ec6baef52edea591eab1a070ed4e18b80d094 build --target=aarch64-apple-darwin
```

And cross-compile easily. Now we just need to wait for Mark-Simulacrum's PRs to get merged and I'll rebase and move the changes to their own regular builders.

**Comment by shepmaster** [comment]
> wait for Mark-Simulacrum's PRs to get merged

This has happened.

> I'll rebase and move the changes to their own regular builders.

As has this.

**Review by emilyalbini — APPROVED** [review]

**Comment by pietroalbini** [comment]
This PR is promoting the `aarch64-apple-darwin` target from Tier 3 to a Tier 2 development platform:

* For infrastructure approval I'd say it doesn't need consensus from the whole team, as we definitely have capacity for a standard sized builder (we don't have XL macOS builders).
* For compiler team approval, I'm nominating this PR.

@rustbot modify labels: T-compiler I-nominated

**Comment by pietroalbini** [comment]
This PR also needs to update `src/doc/rustc/src/platform-support.md`.

**Comment by wesleywiser** [comment]
@rfcbot fcp merge

**Review by emilyalbini — APPROVED** [review]

**Comment by glandium** [comment]
What's missing to have the rust-std artifacts available to `rustup target add`? (it's currently saying the toolchain doesn't contain component 'rust-std' for target 'aarch64-apple-darwin' on the nightly off 8dae8cdcc which is after this was merged)

**Comment by shepmaster** [comment]
https://github.com/rust-lang/rust/pull/77895

As for most everything else for this project, the details [have been discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/aarch64-apple-darwin.20not.20in.20nightly.20manifest).

**Pull request merged**
